src/ctap/pin_protocol_v1.rs

```diff
@@ -510,22 +511,47 @@ impl PinProtocolV1 {
         }
     }
 
-    pub fn has_no_permission_rp_id(&self) -> Result<(), Ctap2StatusCode> {
+    /// Check if no RP ID is associated with the token permission.
+    pub fn has_no_rp_id_permission(&self) -> Result<(), Ctap2StatusCode> {
         if self.permissions_rp_id.is_some() {
             return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
         }
         Ok(())
     }
 
-    pub fn has_permission_for_rp_id(&mut self, rp_id: &str) -> Result<(), Ctap2StatusCode> {
-        if let Some(permissions_rp_id) = &self.permissions_rp_id {
-            if rp_id != permissions_rp_id {
-                return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+    /// Check if no or the passed RP ID is associated with the token permission.
+    pub fn has_no_or_rp_id_permission(&mut self, rp_id: &str) -> Result<(), Ctap2StatusCode> {
+        match &self.permissions_rp_id {
+            Some(p) if rp_id != p => Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID),
+            _ => Ok(()),
+        }
+    }
+
+    /// Check if no RP ID is associated with the token permission, or it matches the hash.
+    pub fn has_no_or_rp_id_hash_permission(
+        &self,
+        rp_id_hash: &[u8],
+    ) -> Result<(), Ctap2StatusCode> {
+        match &self.permissions_rp_id {
+            Some(p) if rp_id_hash != Sha256::hash(p.as_bytes()) => {
+                Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
             }
-        } else {
-            self.permissions_rp_id = Some(String::from(rp_id));
+            _ => Ok(()),
+        }
+    }
+
+    /// Check if the passed RP ID is associated with the token permission.
+    ///
+    /// If no RP ID is associated, associate the passed RP ID as a side effect.
+    pub fn require_rp_id_permission(&mut self, rp_id: &str) -> Result<(), Ctap2StatusCode> {
```

Comment:
Not very happy about the function names here. `has_` functions makes me think they will return `bool` whereas `require_` makes me think about assertions and I expect then to have `Result` as a return value. But I'm not expecting a `require_` function to alter the object.
I know it's not easy but could we try finding better names?

Reply:
I called it `require_` because it is like `has_`, but also sets the RP ID if not present. How about...
- `set_`, but it can't actually overwrite
- `check_`, but it also sets
- `check_and_set_` is a bit long?

If we settle with `check_and_set_`, then `has_` would become `check_`?
To defend the `has_` choice, it is kind of a `bool` with `?` support.

Reply:
I usually call those functions `ensure_` in my personal projects. It ensures the property holds as post-condition regardless of pre-condition. So it may mutate stuff to make it hold. But if it already holds it does nothing, just checks it holds.

Reply:
It seems there's even [a crate](https://docs.rs/ensure/0.3.1/ensure/) that agrees with Julien.
I'm fine leaving tha `has_*` functions as is and change `require_` for `ensure_`.
Is that also good for you Fabian?

Reply:
Very good!